force-app/main/service/DonationHistoryService.cls

```diff
@@ -0,0 +1,176 @@
+        List<DonationHistoryController.DonationHistory> partialSoftCredit = verifyPartialSoftCreditCondition(donationsId, donations);
+        List<DonationHistoryController.DonationHistory> partialAndSoftCredit = verifyPartialAndSoftCreditCondition(softCredit, partialSoftCredit);
+        
+        donationHistory.addAll(hardCredit);
+        donationHistory.addAll(partialAndSoftCredit);
+
+        return donationHistory;
+    }
+    
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description Verify hard credit condition.
+    */
+    private static List<DonationHistoryController.DonationHistory> verifyHardCreditCondition(List<OpportunityContactRole> donations){
+        List<DonationHistoryController.DonationHistory> hardCreditList = new List<DonationHistoryController.DonationHistory>();
+        for(OpportunityContactRole ocr: donations){
+            if (ocr.isPrimary) {
+                DonationHistoryController.DonationHistory hardCredit = new DonationHistoryController.DonationHistory(ocr, HARD_CREDIT);
+                hardCreditList.add(hardCredit);
+            }
+        }
+        return hardCreditList;
+    }
+    
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description Verify soft credit condition.
+    */
+    private static List<DonationHistoryController.DonationHistory> verifySoftCreditCondition(List<OpportunityContactRole> donations){
+        //with the list of donations verify if the Role match the filter rule value on the NPSP_ClosedWon_Opps_SC filter group
+        Filter_Rule__mdt roles = [SELECT Constant__c FROM Filter_Rule__mdt WHERE QualifiedApiName = 'NPSP_OppsSoftCreditRoles'];
+        List<String> softCreditRoles = roles.Constant__c.split(';');
+        List<DonationHistoryController.DonationHistory> softCreditList = new List<DonationHistoryController.DonationHistory>();
+        for(OpportunityContactRole ocr: donations){
+            if (softCreditRoles.contains(ocr.Role) && ocr.isPrimary==false) {
+                DonationHistoryController.DonationHistory softCredit = new DonationHistoryController.DonationHistory(ocr, SOFT_CREDIT);
+                softCreditList.add(softCredit);
+            }
+        }
+        return softCreditList;
+    }
+
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description Filter for ids already classified as Soft Credit.
+    */
+    private static List<Id> filterVerifiedIds(List<DonationHistoryController.DonationHistory> softCredit){
+        List<Id> donationsId = new List<Id>();
+        for(DonationHistoryController.DonationHistory dh: softCredit){
+            donationsId.add(dh.id);
+        }
+        return donationsId;
+    }
+
+    /** 
+    * @author Salesforce.org
+    * @date 2021
+    * @description Verify partial soft credit condition.
+    */
+    private static List<DonationHistoryController.DonationHistory> verifyPartialSoftCreditCondition(List<Id> donationsId, List<OpportunityContactRole> donations){
```

Comment:
Similar comment to the above with regards to renaming `donations` to `credits`.
Similar situation with `donationsId` here. I believe this should be `softCreditIds`. Using `donationIds` here would, at a glance, lead me to think of Opportunities.

Reply:
Done